controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/InstanceList.java

```diff
@@ -36,14 +36,19 @@ private InstanceList(Collection<? extends ApplicationId> items, boolean negate,
     }
 
     /**
-     * Returns the subset of instances where all production deployments are compatible with the given version.
+     * Returns the subset of instances where all production deployments are compatible with the given version,
+     * and at least one known build is compatible with the given version.
      *
      * @param platform the version which applications returned are compatible with
      */
     public InstanceList compatibleWithPlatform(Version platform, Function<ApplicationId, VersionCompatibility> compatibility) {
-        return matching(id -> instance(id).productionDeployments().values().stream()
-                                          .flatMap(deployment -> application(id).revisions().get(deployment.revision()).compileVersion().stream())
-                                          .noneMatch(version -> compatibility.apply(id).refuse(platform, version)));
+        return matching(id ->    instance(id).productionDeployments().values().stream()
+                                             .flatMap(deployment -> application(id).revisions().get(deployment.revision()).compileVersion().stream())
+                                             .noneMatch(version -> compatibility.apply(id).refuse(platform, version))
+                              && application(id).revisions().production().stream()
+                                                .anyMatch(revision -> revision.compileVersion()
+                                                                              .map(compiled -> compatibility.apply(id).accept(platform, compiled))
+                                                                              .orElse(true)));
```

Comment:
... lest apps built against, e.g., major 8, will be eligible for "upgrade" to 7, as long as they have _no_ deployments at all ...